src/server/frontend_wayland/text_input_v3.cpp

```diff
@@ -292,6 +322,26 @@ void mf::TextInputV3::send_text_change(ms::TextInputChange const& change)
         // We are no longer enabled or we don't have a valid serial
         return;
     }
+    if (change.keysym)
+    {
+        auto nano = std::chrono::milliseconds{change.keysym->time};
+        keyboard_device->if_started_then([&](input::InputSink* sink, input::EventBuilder* builder)
+            {
+                auto state = change.keysym->state;
+                auto scan_code = change.keysym->sym;
+                switch (scan_code)
+                {
+                    case XKB_KEY_Return:
+                        // TODO: kwin_wayland maps this to 28, but I am unsure how they get there
+                        scan_code = 28;
```

Comment:
So, (one possible) scancode for Enter is `0x1C`, which *is* 28. Perhaps kwin is attempting to canonicalise the scancode for some reason? There's probably some libxkb function to do whatever this is. :woman_shrugging:

Reply:
Well, to avoid a magic number, there's `KEY_ENTER` from `linux/input-event-codes.h`. (That is, as Chris says, `0x1c`.)

It isn't obvious why this mapping is necessary, but if it works...

I thought I'd seen something similar in the X11 input platform, but all I found was `xcb_keycode_get_scan_code()` which is different "magic".

I don't see an `xkb_*` function to do this mapping.

Reply:
Oh! I wonder if kwin feeds these scancodes into it's xkb state to get a keysym and then does the reverse mapping with `keycode_get_scancode`. It's not really obvious what the protocol is expecting to happen here 🤷‍♀️

Reply:
I assume something similar is required for BkSp?

I don't see Tab, Home, End, Insert, Del, Pg{Up,Down}, arrow keys, meta keys or function keys on the default maliit-keyboard so I suspect there are other mappings required here that we're yet to discover.

Reply:
Funny enough, backspace works without any additional mapping

Reply:
It was time to read `kwin`'s code I suppose :smile: Looks like they do the following in `https://github.com/KDE/kwin/blob/master/src/inputmethod.cpp`:
```cpp
static quint32 keysymToKeycode(quint32 sym)
{
    switch (sym) {
    case XKB_KEY_BackSpace:
        return KEY_BACKSPACE;
    case XKB_KEY_Return:
        return KEY_ENTER;
    case XKB_KEY_Left:
        return KEY_LEFT;
    case XKB_KEY_Right:
        return KEY_RIGHT;
    case XKB_KEY_Up:
        return KEY_UP;
    case XKB_KEY_Down:
        return KEY_DOWN;
    default:
        return KEY_UNKNOWN;
    }
}
```

Reply:
If it's a text input v3, they have no other choice. If it works, it works :shrug: